### PR TITLE
Forms JSON Schema validation

### PR DESCRIPTION
**Pull request by yuval-hazaz**
Create common util function to validate formik forms based on JSON Schema and add schema validation to all forms

### CODE REVIEW

File: packages/amplication-client/src/Application/ApplicationForm.tsx [hunk 1]
```diff
@@ -41,7 +56,14 @@ function ApplicationHome({ app }: Props) {
   const errorMessage = formatError(error);
   return (
     <>
-      <Formik initialValues={app} enableReinitialize onSubmit={handleSubmit}>
+      <Formik
+        initialValues={app}
+        validate={(values: models.App) => {
+          return validate<models.App>(values, FORM_SCHEMA);
+        }}
```
Comment (iddan) [inline packages/amplication-client/src/Application/ApplicationForm.tsx:63]:
```suggestion
        validate={(values: models.App) => validate<models.App>(values, FORM_SCHEMA)}
```

File: packages/amplication-client/src/Application/NewApplication.tsx [hunk 2]
```diff
@@ -74,13 +88,19 @@ const NewApplication = () => {
   }, [history, data]);
 
   return (
-    <Formik initialValues={INITIAL_VALUES} onSubmit={handleSubmit}>
+    <Formik
+      initialValues={INITIAL_VALUES}
+      validate={(values: Values) => {
+        return validate<Values>(values, FORM_SCHEMA);
+      }}
```
Comment (iddan) [inline packages/amplication-client/src/Application/NewApplication.tsx:95]:
```suggestion
      validate={(values: Values) => validate(values, FORM_SCHEMA)}
```

File: packages/amplication-client/src/Entity/EntityForm.tsx [hunk 3]
```diff
@@ -55,6 +74,9 @@ const EntityForm = React.memo(({ entity, applicationId, onSubmit }: Props) => {
     <div className="entity-form">
       <Formik
         initialValues={initialValues}
+        validate={(values: EntityInput) => {
+          return validate<EntityInput>(values, FORM_SCHEMA);
+        }}
```
Comment (iddan) [inline packages/amplication-client/src/Entity/EntityForm.tsx:79]: Same as above

File: packages/amplication-client/src/Entity/NewEntityField.tsx [hunk 4]
```diff
@@ -86,7 +107,14 @@ const NewEntityField = ({ onFieldAdd }: Props) => {
 
   return (
     <>
-      <Formik initialValues={INITIAL_VALUES} onSubmit={handleSubmit}>
+      <Formik
+        initialValues={INITIAL_VALUES}
+        validate={(values: Values) => {
+          return validate<Values>(values, FORM_SCHEMA);
```
Comment (iddan) [inline packages/amplication-client/src/Entity/NewEntityField.tsx:113]: Same as above

File: packages/amplication-client/src/util/formikValidateJsonSchema.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
+import { FormikErrors } from "formik";
+import Ajv from "ajv";
+import { set } from "lodash";
+
+/**
+ * A function to validate a form based on a given JSON Schema.
+ * The function returns a FormikErrors object.
+ * When using formik validation, do no use any DOM validation attributes to avoid the default browser validation and error messages
+ *
+ * Example usage:
```
Comment (iddan) [inline packages/amplication-client/src/util/formikValidateJsonSchema.ts:10]: Use `@example`

File: packages/amplication-client/src/util/formikValidateJsonSchema.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+ * @param validationSchema
+ * The JSON schema for validation
+ *  */
+export function validate<T>(
+  values: T,
+  validationSchema: object
+): FormikErrors<T> {
+  const errors: FormikErrors<T> = {};
+
+  const ajv: Ajv.Ajv = new Ajv({ allErrors: true });
```
Comment (iddan) [inline packages/amplication-client/src/util/formikValidateJsonSchema.ts:31]:
```suggestion
  const ajv = new Ajv({ allErrors: true });
```

File: packages/amplication-client/src/util/formikValidateJsonSchema.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
+import { FormikErrors } from "formik";
+import Ajv from "ajv";
```
Comment (iddan) [inline packages/amplication-client/src/util/formikValidateJsonSchema.ts:2]:
```suggestion
import { Ajv } from "ajv";
```

File: packages/amplication-client/src/util/formikValidateJsonSchema.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+): FormikErrors<T> {
+  const errors: FormikErrors<T> = {};
+
+  const ajv: Ajv.Ajv = new Ajv({ allErrors: true });
+
+  let isValid = ajv.validate(validationSchema, values);
+
+  if (!isValid && ajv.errors) {
+    for (const error of ajv.errors) {
+      const fieldName = error.dataPath.substring(1);
```
Comment (iddan) [inline packages/amplication-client/src/util/formikValidateJsonSchema.ts:37]: What is the format of dataPath. Please add a comment explaining this. 
If you are replacing the `/` I think it's better to use the split string method.

**Review by iddan — COMMENTED** [review]
Good job, this PR includes code unrelated to JSON Schema validation let's make sure it gets merged before we merge this. Please issue my comments

File: packages/amplication-client/src/util/formikValidateJsonSchema.ts (hunk 7 above)
Comment (yuval-hazaz) [reply packages/amplication-client/src/util/formikValidateJsonSchema.ts:2]: It is not working as suggested.

File: packages/amplication-client/src/util/formikValidateJsonSchema.ts (hunk 8 above)
Comment (iddan) [reply packages/amplication-client/src/util/formikValidateJsonSchema.ts:37]: @yuval-hazaz why have you dismissed this comment?

**Review by iddan — APPROVED** [review]

**Pull request merged**